### PR TITLE
Add kraken staking events to the backend

### PR DESCRIPTION
**Pull request by yabirgb**

### CODE REVIEW

File: rotkehlchen/accounting/structures.py [hunk 1]
```diff
@@ -406,3 +408,38 @@ def identifier(self) -> str:
...
+        'receive',
+    ]
+    asset: Asset
+    balance: Balance
+    timestamp: Timestamp
+    location: Location
+
+    @classmethod
+    def from_history_base_entry(cls, event: HistoryBaseEntry) -> 'StakingEvent':
+        event_type = str(event.event_subtype) if event.event_subtype is not None else 'receive'
```
Comment (yabirgb) [inline rotkehlchen/accounting/structures.py:428]: This here is because we forgot to add a subtype to the staking events and rewards are not categorized. We can assume that staking type without subtype is receive something or otherwise include an update in the data migration

File: rotkehlchen/api/v1/resources.py [hunk 2]
```diff
@@ -2075,3 +2076,20 @@ class LimitsCounterResetResource(BaseResource):
...
+    get_schema = StakingQuerySchema
+
+    @use_kwargs(get_schema, location='json_and_query')
+    def get(
+            self,
+            async_query: bool,
+            only_cache: bool,
+            filter_query: StakingQuerySchema,
+    ) -> Response:
+        return self.rest_api.query_kraken_staking_events(
```
Comment (yabirgb) [inline rotkehlchen/api/v1/resources.py:2091]: I noticed that I forgot the kraken name here. Will change after comments. What I wanted to do with this endpoint is to provide something generic so other exchanges/protocols can be queried also using the base history entries

File: rotkehlchen/db/filtering.py [hunk 3]
```diff
@@ -674,6 +675,12 @@ def make(
                 DBStringFilter(and_op=True, column='location_label', value=location_label),
             )
 
+        if location == Location.KRAKEN:
+            if from_ts is not None:
+                from_ts = Timestamp(from_ts * KRAKEN_TS_MULTIPLIER)
+            if to_ts is not None:
+                to_ts = Timestamp(to_ts * KRAKEN_TS_MULTIPLIER)
+
```
Comment (yabirgb) [inline rotkehlchen/db/filtering.py:683]: We have a problem in the history events tables since the precision of the kraken events for the timestamp is higher. We either have to agree on using the same precision for all the events or update them. Otherwise queries by timestamp range in that table don't work properly

File: rotkehlchen/history/events.py [hunk 4]
```diff
@@ -536,3 +538,24 @@ def fail_history_cb(error_msg: str) -> None:
             defi_events,
             ledger_actions,
         )
+
+    def query_kraken_staking_events(
```
Comment (yabirgb) [inline rotkehlchen/history/events.py:542]: I was not sure where to put the function that executes this logic and I went with this one. Also this functions queries the kraken ledger in the full range when the cache is invalidated. I'm not sure this is something you agree with. This makes sense with what we do with trades or deposits. Also if it was recently queried it won't have a big inpact

File: rotkehlchen/tests/api/test_exchanges.py [hunk 5]
```diff
@@ -1165,3 +1165,76 @@ def test_binance_query_pairs(rotkehlchen_api_server_with_exchanges):
...
+            "refid": "XXXXXX",
+            "time": 1640493374.4008,
+            "type": "staking",
+            "subtype": "",
+            "aclass": "currency",
+            "asset": "ETH2",
+            "amount": "0.0000538620",
+            "fee": "0.0000000000",
+            "balance": "0.0003349820"
+        },
```
Comment (yabirgb) [inline rotkehlchen/tests/api/test_exchanges.py:1188]: Took this examples from my ledger

File: rotkehlchen/api/server.py [hunk 6]
```diff
@@ -268,6 +269,7 @@
     ('/database/info', DatabaseInfoResource),
     ('/database/backups', DatabaseBackupsResource),
     ('/locations/associated', AssociatedLocations),
+    ('/staking/', StakingResource),
```
Comment (LefterisJP) [inline rotkehlchen/api/server.py:272]: I am not sure having a generic staking resource is a good idea. We already have staking information for say ETH2. Also AdEx. Why are they not here? You could just for now do `/staking/kraken` and if at some point we add more that can be generic enough we can also use the `/staking` endpoint alone.

File: rotkehlchen/db/filtering.py (hunk 3 above)
Comment (LefterisJP) [reply rotkehlchen/db/filtering.py:683]: yeah I know. I think going with the highest precision available (so kraken's is fine). As long as we deal with ints of the same precision there should be no difference.

File: rotkehlchen/history/events.py [hunk 7]
```diff
@@ -536,3 +538,24 @@ def fail_history_cb(error_msg: str) -> None:
...
+    ) -> List[StakingEvent]:
+        """Queries staking events for kraken.
+        If only_cache is given then only trades cached in the DB are returned.
+        No service is queried.
+        """
+        exchanges_list = self.exchange_manager.connected_exchanges.get(Location.KRAKEN)
+        if exchanges_list is None:
+            return []
+
+        if only_cache is not True:
```
Comment (LefterisJP) [inline rotkehlchen/history/events.py:555]:
```suggestion
        if only_cache is False:
```
giphy

File: rotkehlchen/history/events.py [hunk 8]
```diff
@@ -536,3 +538,24 @@ def fail_history_cb(error_msg: str) -> None:
...
+            for kraken_instance in exchanges_list:
+                kraken_instance.query_kraken_ledgers(start_ts=Timestamp(0), end_ts=ts_now())  # type: ignore  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/history/events.py:557]: no. Extract start_ts and end_ts from the filter query like we do in other calls.

File: rotkehlchen/history/events.py (hunk 4 above)
Comment (LefterisJP) [reply rotkehlchen/history/events.py:542]: This should not be here. You are mixing things here and breaking down well established barriers between modules.

This should be a method of the Kraken object. The exchanges list logic should be in the api/rest.py method.

File: rotkehlchen/history/events.py [hunk 9]
```diff
@@ -536,3 +538,24 @@ def fail_history_cb(error_msg: str) -> None:
...
         )
+
+    def query_kraken_staking_events(
+            self,
+            filter_query: HistoryEventFilterQuery,
+            only_cache: bool,
+    ) -> List[StakingEvent]:
+        """Queries staking events for kraken.
+        If only_cache is given then only trades cached in the DB are returned.
+        No service is queried.
```
Comment (LefterisJP) [inline rotkehlchen/history/events.py:549]:
```suggestion
```

Kraken ledgers are queried.

File: rotkehlchen/tests/api/test_exchanges.py [hunk 10]
```diff
@@ -1165,3 +1165,76 @@ def test_binance_query_pairs(rotkehlchen_api_server_with_exchanges):
     assert some_pairs.issubset(result)
     assert 'FTTBNB' not in result
     assert binance_pairs_num > binanceus_pairs_num
+
+
+@pytest.mark.parametrize('added_exchanges', [(Location.KRAKEN,)])
+def test_kraken_staking(rotkehlchen_api_server_with_exchanges):
+    """Test that the binance endpoint returns some market pairs"""
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_exchanges.py:1172]: steve-harvey-huh

**Review by LefterisJP — COMMENTED** [review]
603dsx

File: rotkehlchen/api/server.py (hunk 6 above)
Comment (yabirgb) [reply rotkehlchen/api/server.py:272]: I went with staking only because this endpoint takes an argument `location` and _should_ work for every staking protocol that has infomration in the `history_events` table. If you prefer I can change it to be `/staking/kraken` only

File: rotkehlchen/history/events.py (hunk 7 above)
Comment (yabirgb) [reply rotkehlchen/history/events.py:555]: https://github.com/rotki/rotki/blame/dc6cb172a726f4af4f114c32aa0f014a84479736/rotkehlchen/history/events.py#L295 :clown_face:

File: rotkehlchen/history/events.py (hunk 8 above)
Comment (yabirgb) [reply rotkehlchen/history/events.py:557]: Didn't know this existed. Thanks!

File: rotkehlchen/api/rest.py [hunk 11]
```diff
@@ -3906,3 +3907,48 @@ def get_associated_locations(self) -> Response:
...
+    ) -> Dict[str, Any]:
+        exchanges_list = self.rotkehlchen.exchange_manager.connected_exchanges.get(
+            Location.KRAKEN,
+        )
+        if exchanges_list is None:
+            events = []
+        else:
+            if only_cache is False:
+                for kraken_instance in exchanges_list:
+                    kraken_instance.query_kraken_ledgers(   # type: ignore
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:3924]: Can't this raise? If it can it needs to be checked.

Also from what I see perhaps we catch all in there and return False in case of error. That `False` needs to be checked for an an appropriate API error returned.

File: rotkehlchen/api/v1/resources.py (hunk 2 above)
Comment (LefterisJP) [reply rotkehlchen/api/v1/resources.py:2091]: I am not sure what you mean by this comment

File: docs/api.rst [hunk 12]
```diff
@@ -10126,3 +10126,82 @@ Get associated locations
...
+                  "amount": "0.0000103220",
+                  "usd_value": "0"
+              },
+              {
+                  "event_type": "staking receive asset",
+                  "asset": "ETH2",
+                  "timestamp": 16367401989674,
+                  "location": "kraken",
+                  "amount": "0.0600000000",
+                  "usd_value": "0"
```
Comment (LefterisJP) [inline docs/api.rst:10185]: It just occured to me that at the result of this call we want to have usd_values. Otherwise it's sort of useless. Do you think it would make sense to do a run of the "usd detection" logic for the stake events during this api call? Only run it ofc for stake events and only for stake events that don't already have a value.

File: rotkehlchen/api/v1/resources.py (hunk 2 above)
Comment (yabirgb) [reply rotkehlchen/api/v1/resources.py:2091]: Nothing now. Is what about making sense between the route name and the name of the methods used

File: rotkehlchen/api/rest.py (hunk 11 above)
Comment (yabirgb) [reply rotkehlchen/api/rest.py]: I've added a message so now the frontend can know if the API failed and returned only database entries. Also during the testing with real data saw that a catch in the kraken class was missing for remoterror. This function shouldn't raise any error.

File: rotkehlchen/externalapis/cryptocompare.py [hunk 13]
```diff
@@ -843,6 +843,13 @@ def query_historical_price(
             )
 
         log.debug('Got historical price from cryptocompare', from_asset=from_asset, to_asset=to_asset, timestamp=timestamp, price=price)  # noqa: E501
+        GlobalDBHandler().add_historical_prices(entries=[HistoricalPrice(
+            from_asset=from_asset,
+            to_asset=to_asset,
+            source=HistoricalPriceOracle.CRYPTOCOMPARE,
+            timestamp=timestamp,
+            price=price,
+        )])
```
Comment (yabirgb) [inline rotkehlchen/externalapis/cryptocompare.py:852]: Added this here to save prices from cryptocompare queried using this method. For what I saw it was missing

Comment (LefterisJP) [reply rotkehlchen/externalapis/cryptocompare.py:852]: Hmm should be fine. I think we just did cache in the sense of searching all historical queries in histohour before, but this should also work fine.

File: rotkehlchen/rotkehlchen.py [hunk 14]
```diff
@@ -903,3 +910,44 @@ def create_oracle_cache(
...
+                # Shouldn't raise DeserializationError since data comes from the database
+                amount = deserialize_fval(
+                    value=amount,
+                    name='historic base entry usd_value query',
+                    location='query_missing_prices',
+                )
+                price = inquirer.query_historical_price(
+                    from_asset=Asset(asset_name),
+                    to_asset=A_USD,
+                    timestamp=Timestamp(int(timestamp / KRAKEN_TS_MULTIPLIER)),
```
Comment (LefterisJP) [inline rotkehlchen/rotkehlchen.py:939]: Perhaps we can make sure this timestamp is returned divided and to the "our format" from the `get_rows_missing_prices_in_base_entries()`?

It's good to serialize/deserialize as early as possible to our format. Otherwise we forget where we have what.

File: rotkehlchen/tests/api/test_exchanges.py [hunk 15]
```diff
@@ -1165,3 +1166,89 @@ def test_binance_query_pairs(rotkehlchen_api_server_with_exchanges):
...
+    assert len(result) == 3
+    assert result[0]['event_type'] == 'receive'
+    assert result[1]['event_type'] == 'staking receive asset'
+    assert result[2]['event_type'] == 'staking deposit asset'
+    assert result[0]['asset'] == 'ETH2'
+    assert result[1]['asset'] == 'ETH2'
+    assert result[2]['asset'] == 'ETH'
+    assert result[0]['usd_value'] == '0.219353533620'
+    assert result[1]['usd_value'] == '242.570400000000'
+    assert result[2]['usd_value'] == '-242.570400000000'
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_exchanges.py:1254]: The coment I made about negative values in the docs applies here too

File: rotkehlchen/tests/api/test_history.py [hunk 16]
```diff
@@ -220,8 +220,8 @@ def test_query_history_remote_errors(rotkehlchen_api_server_with_exchanges):
     warnings = rotki.msg_aggregator.consume_warnings()
     assert len(warnings) == 0
     errors = rotki.msg_aggregator.consume_errors()
-    assert len(errors) == 3
-    assert 'Etherscan API request http://someurl.com returned invalid JSON response: [{' in errors[0]  # noqa: E501
+    assert len(errors) == 5
+    assert 'Etherscan API request http://someurl.com returned invalid JSON response: [{' in errors[2]  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_history.py:224]: Hm ... what are errors[0] and errors[1] here now? And why are you not testing for them?

Comment (yabirgb) [reply rotkehlchen/tests/api/test_history.py:224]: They are error related to the kraken api calls. We were missing a remote error catch and I added it with an error message

File: rotkehlchen/api/rest.py [hunk 17]
```diff
@@ -3906,3 +3908,53 @@ def get_associated_locations(self) -> Response:
...
+            Location.KRAKEN,
+        )
+        message = ''
+        if exchanges_list is None:
+            events = []
+        else:
+            if only_cache is False:
+                for kraken_instance in exchanges_list:
+                    with_errors = kraken_instance.query_kraken_ledgers(   # type: ignore
+                        start_ts=filter_query.from_ts / KRAKEN_TS_MULTIPLIER,
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/api/rest.py:3927]: That's the issue that I commented before. Since the entries in the table queried have the higher resolution timestamps they filter have them adjusted to that resolution. Every other places that requires having a lower resolution needs to properly scale them.

If I change this I believe I would need to modify the prepare method of the filters

File: rotkehlchen/rotkehlchen.py [hunk 18]
```diff
@@ -903,3 +910,44 @@ def create_oracle_cache(
...
         self.cryptocompare.create_cache(from_asset, to_asset, purge_old)
+
+    def query_missing_prices(
+        self,
+        filter_query: HistoryEventFilterQuery,
+    ) -> None:
+        """Queries missing prices for HistoryBaseEntry in database based
+        on the query filter used"""
+        # Use a deepcopy to avoid mutations in the filter query if it is used later
+        new_filter_query = copy.deepcopy(filter_query)
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/rotkehlchen.py:921]: Yes, this filter has pagination and is later used in the logic (`rest.py`) to query the filtered events. What I do is update missing prices for events that will be returned

File: rotkehlchen/api/rest.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/api/rest.py:3927]: Ah I see. Hmmm I think these conversions would make sense to be happening in the filter class eventually

File: rotkehlchen/rotkehlchen.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/rotkehlchen.py:921]: Aha I see, so you don't query prices for all events but only for this page's events? Hmmm this will end up being weird at page switch. We really need to have a task going through all of them and getting prices if possible.

Essentially this: https://github.com/rotki/rotki/issues/3865

**Comment by yabirgb** [comment]
What I did in the last commit was:
- Adding the recurring task to fetch prices
- Improve how we handle the tiemestamp resolution

File: rotkehlchen/rotkehlchen.py [hunk 19]
```diff
@@ -324,6 +329,13 @@ def unlock_user(
             batch_size=ICONS_BATCH_SIZE,
             sleep_time_secs=ICONS_QUERY_SLEEP,
         )
+        self.greenlet_manager.spawn_and_track(
```
Comment (LefterisJP) [inline rotkehlchen/rotkehlchen.py:332]: no no no not here. This should be as we do it in tasks/manager.py

File: rotkehlchen/rotkehlchen.py [hunk 20]
```diff
@@ -903,3 +915,38 @@ def create_oracle_cache(
...
+
+    def periodically_query_history_events_prices(self, sleep_time_secs: int) -> None:
+        """
+        Query periodically the history_events table to find entries missing a proper
+        usd_value column value.
+        """
+        query = HistoryEventFilterQuery.make()
+        while True:
+            self.query_missing_prices(filter_query=query)
+            gevent.sleep(sleep_time_secs)
```
Comment (LefterisJP) [inline rotkehlchen/rotkehlchen.py:952]: And no absolutely not like that.

This will be a task that runs forever and ever if even 1 price for 1 asset at a single timestamp can not be found. Checks tasks/manager.py

File: rotkehlchen/tests/api/test_history.py [hunk 21]
```diff
@@ -220,8 +220,10 @@ def test_query_history_remote_errors(rotkehlchen_api_server_with_exchanges):
     warnings = rotki.msg_aggregator.consume_warnings()
     assert len(warnings) == 0
     errors = rotki.msg_aggregator.consume_errors()
-    assert len(errors) == 3
-    assert 'Etherscan API request http://someurl.com returned invalid JSON response: [{' in errors[0]  # noqa: E501
+    assert len(errors) == 5
+    assert 'kraken' in errors[0]
+    assert 'kraken' in errors[1]
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_history.py:225]:
```suggestion
    assert all('kraken' in e for e in errors[:2])
```

File: rotkehlchen/api/rest.py [hunk 22]
```diff
@@ -3906,3 +3908,58 @@ def get_associated_locations(self) -> Response:
...
+                    with_errors = kraken_instance.query_kraken_ledgers(   # type: ignore
+                        start_ts=filter_query.from_ts,
+                        end_ts=filter_query.to_ts,
+                    )
+                    if with_errors:
+                        kraken_names.append(kraken_instance.name)
+                if len(kraken_names) != 0:
+                    message = f'Failed to query some events from Kraken exchanges {",".join(kraken_names)}'  # noqa: E501
+
+            # After 3865 we should have a recurring tasks that queries for missing prices but
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:3936]: Don't you essentially implement 3865 at the last commit?

File: rotkehlchen/tasks/manager.py [hunk 23]
```diff
@@ -279,6 +282,19 @@ def _maybe_schedule_exchange_history_query(self) -> None:
...
         self.last_exchange_query_ts[exchange.location_id()] = now
 
+    def _maybe_query_missing_prices(self) -> None:
+        query = HistoryEventFilterQuery.make()
+        task_name = 'Periodically query history events prices'
+        log.debug(f'Scheduling task to {task_name}')
+        self.greenlet_manager.spawn_and_track(
+            after_seconds=None,
+            task_name=task_name,
+            exception_is_error=True,
```
Comment (LefterisJP) [inline rotkehlchen/tasks/manager.py:292]: I would advise here, just like with other tasks to limit it a bit. It can become a really long runnig tasks if there is say 1000 entries without a price.

Also there can be entries that do not have a price and there is no way to find one with the current oracles. So the scheduler will re-schedule, and re-schedule, and re-schedule, and re-schedule, and re-schedule .... you get the drift.

For example we face this problem where querying coingecko for icons. Which is why we keep the ids of assets for which we could not find an icon in memory so they are not scheduled again: https://github.com/rotki/rotki/blob/32a9010576023cf69c6202731480d9b00136e966/rotkehlchen/icons.py#L41
https://github.com/rotki/rotki/blob/32a9010576023cf69c6202731480d9b00136e966/rotkehlchen/icons.py#L88
https://github.com/rotki/rotki/blob/32a9010576023cf69c6202731480d9b00136e966/rotkehlchen/icons.py#L164-L166

File: rotkehlchen/api/rest.py (hunk 22 above)
Comment (yabirgb) [reply rotkehlchen/api/rest.py:3936]: Yes! The comment is to let know the reader that after implementing 3865 we have a mechanism to query the prices but we ensure that the retrieve page has prices queried by calling the logic for missing prices

Comment (LefterisJP) [reply rotkehlchen/api/rest.py:3936]: yeah so just say that there is a recurring task (notice singular) that queries for missing prices.

"We should have" implies it will happen in the future

File: rotkehlchen/api/rest.py [hunk 24]
```diff
@@ -3906,3 +3911,73 @@ def get_associated_locations(self) -> Response:
...
+                    with_errors = kraken_instance.query_kraken_ledgers(   # type: ignore
+                        start_ts=filter_query.from_ts,
+                        end_ts=filter_query.to_ts,
+                    )
+                    if with_errors:
+                        kraken_names.append(kraken_instance.name)
+                if len(kraken_names) != 0:
+                    message = f'Failed to query some events from Kraken exchanges {",".join(kraken_names)}'  # noqa: E501
+
+            # After 3865 we have a recurring tasks that queries for missing prices but
```
Comment (LefterisJP) [inline rotkehlchen/api/rest.py:3939]:
```suggestion
            # After 3865 we have a recurring task that queries for missing prices but
```

I specifically wrote "singular" not plural in my last comment on this https://github.com/rotki/rotki/pull/3904#discussion_r781311720

File: rotkehlchen/db/dbhandler.py [hunk 25]
```diff
@@ -3521,3 +3523,22 @@ def get_history_events(self, filter_query: HistoryEventFilterQuery) -> List[Hist
...
+    def iterate_rows_missing_prices_in_base_entries(
+        self,
+        filter_query: HistoryEventFilterQuery,
+    ) -> Tuple[Iterator[Tuple[str, FVal, Asset, Timestamp]], int]:
+        """
+        Get missing prices for history base entries based on filter query
+        """
+        query, bindings = filter_query.prepare()
+        main_query = 'SELECT identifier, amount, asset, timestamp FROM history_events ' + query
+        print(main_query, bindings)
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:3536]: no

```suggestion
```

File: rotkehlchen/tasks/history_base_entries.py [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+    entries_missing_prices, count = db.iterate_rows_missing_prices_in_base_entries(
+        filter_query=new_filter_query,
+    )
+    return entries_missing_prices, count
+
+
+def query_missing_prices(
+    rows: Iterator[Tuple[str, FVal, Asset, Timestamp]],
+    db: 'DBHandler',
+    ignore_list: Set[str],
```
Comment (LefterisJP) [inline rotkehlchen/tasks/history_base_entries.py:51]: Sorry but no Yabir, this entire code with doing all these extra gymnastics just to keep the iterator is terribly ugly and unnecessarily complicated. I told you in discord to not do this.

Since I last checked you seem to have:

1. Added a new exclude filter to the DB query
2. Messed up ignore_list with ignore_set. Decide. Is it a list? Is it a set? Treat it as one all across the code.
3. You are passing a set argument through multiple functions many levels deep, turning it into a list some times and then mutate it. Mutating an argument is a terrible, terrible practise. It's hard to follow, you have no guarantees on the code, you have to know the argument is mutated and you will forget that it does so in a week. Imagine someone new reading the code.

All this for what? To just keep returning an iterator? Please keep it simple, keep it readable and keep proper coding interfaces and let's close this PR.

Please do the following:
1. Query X history events with 0 value. I already told you to limit the number of event queried in a single task. Look at what we do for transaction receipts.
2. For each of the X history events query the price. If the price query fails note it in the task manager's state. Or in whatever object's state. This is where you have to keep proper interfaces and abstractions. Passing a set to be mutated by some random functions is BAD, BAD, BAD, BAD. A developer should not have to read the code to kind of guess where things should happen. Keep proper abstractions and interfaces.
3. If the price query succeeds save it in the DB

**Review by LefterisJP — COMMENTED** [review]
still needs a big change in the way the task manager price query for historic events happen

File: rotkehlchen/db/dbhandler.py [hunk 27]
```diff
@@ -3521,3 +3524,45 @@ def get_history_events(self, filter_query: HistoryEventFilterQuery) -> List[Hist
...
+        """
+        Get missing prices for history base entries based on filter query
+        """
+        query, bindings = filter_query.prepare()
+        query = 'SELECT identifier, amount, asset, timestamp FROM history_events ' + query
+        cursor = self.conn.cursor()
+        result = []
+        for identifier, amount_raw, asset_name, timestamp in cursor.execute(query, bindings):
+            if identifier in ignored_ids:
+                continue
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:3542]: Few things off here.

1. Do not do in cursor.execute(). Separate concerns please.
2. Why do the identifier check outside of sqlite? Keep that filter you had in the previous partof the PR. Do not pass ignored_ids here but just pass the proper filter including ignored ids.
3. 
```suggestion
        query = 'SELECT identifier, amount, asset, timestamp FROM history_events ' + query
        result = []
        cursor = self.conn.cursor()
        cursor.execute(query, bindings)
        for identifier, amount_raw, asset_name, timestamp in cursor:
```

File: rotkehlchen/tasks/manager.py [hunk 28]
```diff
@@ -279,6 +287,70 @@ def _maybe_schedule_exchange_history_query(self) -> None:
...
+                method=self.query_missing_prices_of_base_entries,
+                entries_missing_prices=entries,
+            )
+
+    def get_base_entries_missing_prices(
+        self,
+        query_filter: HistoryEventFilterQuery,
+    ) -> List[Tuple[str, FVal, Asset, Timestamp]]:
+        """
+        Searches base entries missing usd prices that have not previosly been checked in
```
Comment (LefterisJP) [inline rotkehlchen/tasks/manager.py:309]:
```suggestion
        Searches base entries missing usd prices that have not previously been checked in
```

File: rotkehlchen/tasks/manager.py [hunk 29]
```diff
@@ -279,6 +287,70 @@ def _maybe_schedule_exchange_history_query(self) -> None:
...
+                    timestamp=timestamp,
+                )
+            except (NoPriceForGivenTimestamp, RemoteError) as e:
+                log.error(
+                    f'Failed to find price for {asset} at {timestamp} in base '
+                    f'entry {identifier}. {str(e)}.',
+                )
+                self.base_entries_ignore_set.add(identifier)
+                continue
+            usd_value = amount * price
```
Comment (LefterisJP) [inline rotkehlchen/tasks/manager.py:346]: nitpick: This helps separate sections

```suggestion
                continue

            usd_value = amount * price
```

**Review by LefterisJP — COMMENTED** [review]
Some changes

File: rotkehlchen/db/filtering.py [hunk 30]
```diff
@@ -333,6 +342,17 @@ def prepare(self) -> Tuple[List[str], List[Any]]:
...
 
+@dataclass(init=True, repr=True, eq=True, order=False, unsafe_hash=False, frozen=False)
+class DBIgnoreValuesFilter(DBFilter):
+    column: str
+    values: List[Any]
+
+    def prepare(self) -> Tuple[List[str], List[Any]]:
+        if len(self.values) == 0:
+            return [], []
+        return [f'{self.column} NOT IN ({", ".join(["?"] * len(self.values))})'], [self.values]
```
Comment (LefterisJP) [inline rotkehlchen/db/filtering.py:353]: This is tricky. Probably for now we can merge as is but this is incomplete and can hide a bug.

1. Don't use `Any` as type here. Use as exact types as you can and we know that this is only used for strings here so use List[str].
2.  If the list ever ends up containing a `None` this will fail very sneakly. https://stackoverflow.com/questions/129077/null-values-inside-not-in-clause
Null checks don't work with IN. If we ever need a null check here (say the list can have `None`) then it should be added as an addition OR self.column IS NULL or IS NOT NULL depending on query

**Review by LefterisJP — APPROVED** [review]
An important note left but believe we can merge for now and improve later.

Failed tests are due to the AVAX changes.

File: rotkehlchen/db/filtering.py (hunk 30 above)
Comment (yabirgb) [reply rotkehlchen/db/filtering.py:353]: oh okey. I was using str in the previous commit but wanted to generalize. From now on I'll go restrictive and expand types as we need it
